[MRG] Include pxd-files into the installation

realead opened the pull request:
#### Reference Issues/PRs

Fixes #14847 


#### What does this implement/fix? Explain your changes.

All pxds are now copied to the installation, so they can be reused for building Cython-extension by the user.

`pyximport` is used to compile the test-pyx-files during the runtime. The rebuild is forced so the cache isn't used.


#### Any other comments?

I didn't include pxd-files from sklearn.svm (libsvm.pxd and liblinear.pxd), because in order to build them not only the h-files from `src`-folder are needed, but also c and cpp files, because of this [include](https://github.com/scikit-learn/scikit-learn/blob/cad0fb434f4a199b563948325197f1e11af83c92/sklearn/svm/libsvm.pxd#L42):

```
cdef extern from "libsvm_helper.c":
   ....
```
which leaks all definitions into the cythonized cpp-file, so the linker needs everything from cpp in order to be able to resolve all those leaked symbols.

If these pxd files should be included into the installation as well, first one needs to introduce libsvm_helper.h and build libsvm_helper.c as a source file and  be including it into the pxd (and the same for liblinear.pxd)

realead committed:
adding pxd-files from neighbors to the installation. Adding tests to
test, that pxd files are in the installation and can be cimported. For
that we utilize pyximport and force rebuild every time, thus the cache
is not used

realead committed:
adding pxd-files from tree to the installation.

realead committed:
adding pxd-files from ensemble to the installation.

realead committed:
adding pxd-files from linear_model to the installation.

realead committed:
adding pxd-files from utils to the installation. cpp-pyx files need a special treatment (language-setting)

realead commented:
I have some questions:

   -  pyximport uses ./pyxbld for caching the results of the build. The rebuild is enforced, but still some data stays in the cache (it should be no problem, but you never know) Is this a problem? It is possible to put the build-cache in any other directory which can be wipped after the run. Is there already such a place?

  - [ this code](https://github.com/scikit-learn/scikit-learn/pull/14896/commits/1715927f38218a48ceb583901bddca3b80f79866#diff-45cde2c2674faefc0027030f5a03b044) for building and importing pyx-files via pyximport is used for every submodule. It is probably a good idea to extract it as a test-utility-function. Is there a good place where  such test-suite functionality can be put, so it can be easily reused?

realead committed:
introducing building/loading helper for pyx-files

rth reviewed (commented):
Thanks for looking into this @realead !

It's a thorough approach, but I wonder if we can get away with avoiding compilation tests. In particular, adding the relevant files to `Manifest.in` and then add a single test that checks for the includes pxd somewere in,
```bash
sklearn/tests/test_check_build.py
```
i.e. that the following files exist,
```bash
$ find sklearn -iname '*.pxd'
sklearn/ensemble/_hist_gradient_boosting/common.pxd
sklearn/linear_model/sgd_fast.pxd
sklearn/neighbors/typedefs.pxd
sklearn/neighbors/dist_metrics.pxd
sklearn/neighbors/quad_tree.pxd
sklearn/svm/libsvm.pxd
sklearn/svm/liblinear.pxd
sklearn/tree/_tree.pxd
sklearn/tree/_criterion.pxd
sklearn/tree/_utils.pxd
sklearn/tree/_splitter.pxd
sklearn/utils/murmurhash.pxd
sklearn/utils/weight_vector.pxd
sklearn/utils/_cython_blas.pxd
sklearn/utils/_random.pxd
sklearn/utils/fast_dict.pxd
sklearn/utils/seq_dataset.pxd
```
with respect to
```py
import sklearn

base_dir = sklearn.__file__
```
could be enough. Since in some CI we build wheels and run tests with `pytest --pyargs sklearn`. So that would check that these files are included in wheels. I'm not sure there is a need to check that we can cimport them in Cython?

rth commented:
I agree that dependencies between pxd is a tricky question, but these are integration/build tests that only need to be run once before a release. Making them run as part of the unit tests suite, particularly if they need to compile something each time they are run, is problematic. The resulting `.so` would also add some measurable size to the built wheels.

I would still prefer,
  - either a standalone way to check this, e.g. under `maint_tools`
  - or some lighter tests in the test suite (that don't require compilation)

The next release is happening soon and I think it would be good to have it there. Also because there were some refactoring of the API, the paths to some of the pxd might have changed and it would be really useful to double-check that the inlcuded ones are still correct.

Could you also please merge master in?

realead committed:
moving tests to maint_tools

realead commented:
@rth 

I'm not sure, why this is a problem to build and also don't think that so-files will end up in the wheels (pyximport put them into home/<user_name>/.pyxbld/lib.linux-x86_64-3.7 or whatever OS/Python version). 

However, I have rebased and  introduced/moved tests to a standalone tester under maint_tools.

PS: not sure what this failing test is about, seems to be a missing sh-file -hopefully my PR has nothing to do with it.

realead commented:
no longer needed.
